### PR TITLE
[Feat] AffineTransformation alignment and other fixes

### PR DESCRIPTION
**Pull request by shijianjian**
Something to be done:
- RandomAffine parameter alignment with TorchVision
- Expose interpolation/resample methods #486 
- [POTENTIAL] To have an ```kornia.nn``` module to aggregate all ```nn.Module``` classes.
- [POTENTIAL] Expose ```same_on_batch``` parameter to ```random_apply_*``` functional APIs.

### CODE REVIEW

**Comment by shijianjian** [comment]
I think all parameters now are very much aligned with TorchVision now. But we still have a little difference with TorchVision. For details, you may check it up [here](https://colab.research.google.com/drive/1T20UNAG4SdlE2n2wstuhiewve5Q81VpS#scrollTo=MYe6hFxvp4Ri)

After some digging, I think the issue may be raised by the 'scale' parameter.
Torch calcs, [here](https://pytorch.org/docs/stable/_modules/torchvision/transforms/functional.html):
```python

    # RSS without scaling
    a = cos(rot - sy) / cos(sy)
    b = -cos(rot - sy) * tan(sx) / cos(sy) - sin(rot)
    c = sin(rot - sy) / cos(sy)
    d = -sin(rot - sy) * tan(sx) / cos(sy) + cos(rot)

    # Inverted rotation matrix with scale and shear
    # det([[a, b], [c, d]]) == 1, since det(rotation) = 1 and det(shear) = 1
    M = [d, -b, 0,
         -c, a, 0]
    M = [x / scale for x in M]
```

While we are doing, in ```imgwarp.py```:
```python
    scaled_rotation: torch.Tensor = angle_to_rotation_matrix(angle) * scale.view(-1, 1, 1)
```

**Comment by edgarriba** [comment]
Regarding `kornia.nn` I agree, could be a good idea. Can you open an issue and start the discussion there and if you have already any idea how would you structure inside?

File: kornia/augmentation/random.py [hunk 1]
```diff
@@ -227,12 +227,12 @@ def random_affine_gen(
...
-    transform: torch.Tensor = _get_random_affine_params(
-        batch_size, height, width, degrees_tmp, translate, scale, shear_tmp, same_on_batch)
-    return dict(transform=transform)
+    return _get_random_affine_params(
+        batch_size, height, width, degrees_tmp, translate, scale, shear_tmp, resample, same_on_batch)
 
 
 def random_rotation_gen(batch_size: int, degrees: FloatUnionType,
+                        interpolation: Resample = Resample.BILINEAR,
                         same_on_batch: bool = False) -> Dict[str, torch.Tensor]:
```
[the remark this answers is not in the document]
Comment (shijianjian) [reply kornia/augmentation/random.py:236]: To be honest, I do not think passing to GPU will make it faster since we are not doing intensive computations here. GPU is good at vector computations but please do remember that each core is pretty weak. I did a bit comparison here (torch GPU, torch CPU, numpy).
https://colab.research.google.com/drive/15okeDVpv_4H8dl8X6rzyi6BAcyNOPjbU

It shows that:
- Torch GPU: 400 us
- Torch CPU: 127 us
- NumPy: 30 us.

For the best param gen performance, I think NumPy is much better while it may break Kornia's philosophy.

File: kornia/augmentation/augmentation.py [hunk 2]
```diff
@@ -4,6 +4,7 @@
 import torch.nn as nn
 from torch.nn.functional import pad
 
+from kornia.constants import Resample
```
[the remark this answers is not in the document]
Comment (shijianjian) [reply kornia/augmentation/augmentation.py:7]: Okay. Then I prefer to remove the enum then.

File: kornia/augmentation/random.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (shijianjian) [reply kornia/augmentation/random.py:236]: > I don't think interpolation should be in the signature since it does not affect in the generator

Yes. But it would have a better API consistency to put it in.



> Thinking about about maybe a more clear name might help like `random_rotation_generator`

This makes sense.

**Comment by edgarriba** [comment]
@shijianjian do you have a gpu to test ?

**Comment by shijianjian** [comment]
> @shijianjian do you have a gpu to test ?

I used colab for now.

File: kornia/augmentation/functional.py [hunk 3]
```diff
@@ -659,6 +677,41 @@ def apply_adjust_gamma(input: torch.Tensor, params: Dict[str, torch.Tensor],
...
     if return_transform:
-        raise NotImplementedError
-    return adjust_gamma(input, params['gamma_factor'].type_as(input))
+        identity: torch.Tensor = torch.eye(3, device=input.device, dtype=input.dtype).repeat(input.shape[0], 1, 1)
+        return transformed, identity
+
+    return transformed
+
+
+def _compose_affine_matrix_3x3(translations: torch.Tensor,
```
[the remark this answers is not in the document]
Comment (shijianjian) [reply kornia/augmentation/functional.py:692]: I will move it to ```kornia.geometry.transform.imgwarp.py``` and rename it as ```get_affine_matrix2d``` then?

File: kornia/augmentation/functional.py [hunk 4]
```diff
@@ -517,8 +522,8 @@ def apply_crop(input: torch.Tensor, params: Dict[str, torch.Tensor], return_tran
 
     return crop_by_boxes(
         input,
-        params['src'].type_as(input),
-        params['dst'].type_as(input),
+        params['src'].float(),
+        params['dst'].float(),
```
Comment (edgarriba) [inline kornia/augmentation/functional.py]: I would set it as input.dtype, otherwise gradcheck might fail

File: kornia/augmentation/functional.py [hunk 5]
```diff
@@ -583,7 +588,7 @@ def apply_adjust_brightness(input: torch.Tensor, params: Dict[str, torch.Tensor]
         torch.Tensor: Adjusted image.
     """
     input = _transform_input(input)
-    transformed = adjust_brightness(input, params['brightness_factor'].type_as(input) - 1)
+    transformed = adjust_brightness(input, params['brightness_factor'] - 1)
```
Comment (edgarriba) [inline kornia/augmentation/functional.py:591]: set at least dtype for gradcheck test

File: kornia/constants.py [hunk 6]
```diff
@@ -1,6 +1,9 @@
 from typing import Union, TypeVar
 from enum import Enum
 
+import torch
+
+PI = torch.tensor(3.14159265358979323846)
```
Comment (edgarriba) [inline kornia/constants.py:6]: leave it in undercase, was inspired by numpy here: https://docs.scipy.org/doc/numpy/reference/constants.html#numpy.pi

File: kornia/geometry/transform/crop.py [hunk 7]
```diff
@@ -144,7 +144,8 @@ def center_crop(tensor: torch.Tensor, size: Tuple[int, int], interpolation: str
         [dst_w - 1, dst_h - 1],
         [0, dst_h - 1],
     ]], device=tensor.device).expand(points_src.shape[0], -1, -1)
-    return crop_by_boxes(tensor, points_src, points_dst, interpolation, return_transform=return_transform)
+    return crop_by_boxes(
+        tensor, points_src.float(), points_dst.float(), interpolation, return_transform=return_transform)
```
Comment (edgarriba) [inline kornia/geometry/transform/crop.py:148]: don't force floats, we need to test with double precision with gradcheck

**Comment by shijianjian** [comment]
Hi @edgarriba,

I think it is now finished and ready to be merged.

**Comment by edgarriba** [comment]
@shijianjian let me know when are you ready to merge so that I can start preparing the release by the weekend. One additional comment, for a separated PR, I observed a pattern in each module where the forward call is exactly the same for every single function. I thing this can included in the base class, and same for occurrence probability parameter. This will reduce pretty much the effort for adding new augmentions to just provide the functional `apply_fcn` and `random_generation_fcn`.

**Comment by edgarriba** [comment]
another observed pattern is the `get_params` which mostly rely on input shape, so why not having uniform parameters and pass directly the input. This could be achieved with a static method like `RandomRotation.get_params_as(input)` that will go in concordance with similar pytoch *as functions. This I believe will help also to easily included more functions to the module.

/cc @justusschock do you have anything similar in rising?

**Comment by shijianjian** [comment]
Hi @edgarriba,

I have updated the ```get_param``` and ```forward``` logics. It should be all good now.

**Comment by edgarriba** [comment]
@shijianjian great ! let me know when the code is ready to merge. At least for now looks good to me

**Comment by shijianjian** [comment]
Hi @edgarriba,

I have added tests for the newly introduced ```same_on_batch``` param. This PR should be finalized now. It is alright to be merged.

New functionalities like RandomMotionBlur will be added later, probably not in this release.

----
I will give a benchmark tomorrow, about:
- Op-level comparison between TorchVision and Kornia (batchsize is 1), resulting in a table.
- Op-level comparison of Kornia ops in different batch sizes (batchsize 1, 8, 16, 32, 64), resulting in a line chart (x-axis: batch sizes, y-axis: time-cost)

In the end, I hope the benchmark would encourage people using a larger batchsize since the per-sample speed will be faster.

**Comment by edgarriba** [comment]
@shijianjian the code looks good. Merging this and preparing the release docs.

One thing we should add in docs (will briefly do in release notes) is to give an example about how to create a new operator based on `kornia.augmentations`. We would like people to use this as framework for their accelerated augmentations and make easy the experience to add more transforms (contribute or just for own use). Basically a simple example inheriting from `K.augmentation.BaseAugmenation`.
